src/XMonad/Core.hs

```diff
@@ -447,126 +449,95 @@ runOnWorkspaces job = do
-      Just path -> return path
-      Nothing   -> do
-        primary <- io (head possibles)
-        io (createDirectoryIfMissing True primary)
-        return primary
-
-  where
-    go []     = return Nothing
-    go (x:xs) = do
-      dir    <- io x
-      exists <- io (doesDirectoryExist dir)
-      if exists then return (Just dir) else go xs
-
--- | Simple wrapper around @findFirstDirOf@ that allows the primary
--- path to be specified by an environment variable.
-findFirstDirWithEnv :: MonadIO m => String -> [IO FilePath] -> m FilePath
-findFirstDirWithEnv envName paths = do
-    envPath' <- io (getEnv envName)
-
-    case envPath' of
-      Nothing      -> findFirstDirOf paths
-      Just envPath -> findFirstDirOf (return envPath:paths)
-
--- | Helper function to retrieve the various XDG directories.
--- This has been based on the implementation shipped with GHC version 8.0.1 or
--- higher. Put here to preserve compatibility with older GHC versions.
-getXDGDirectory :: XDGDirectory -> FilePath -> IO FilePath
-getXDGDirectory xdgDir suffix =
-  normalise . (</> suffix) <$>
-  case xdgDir of
-    XDGData   -> get "XDG_DATA_HOME"   ".local/share"
-    XDGConfig -> get "XDG_CONFIG_HOME" ".config"
-    XDGCache  -> get "XDG_CACHE_HOME"  ".cache"
+-- | All the directories that xmonad will use.  They will be used for
+-- the following purposes:
+--
+-- * @dataDir@: This directory is used by XMonad to store data files
+-- such as the run-time state file and the configuration binary
+-- generated by GHC.
+--
+-- * @cfgDir@: This directory is where user configuration files are
+-- stored (e.g, the xmonad.hs file).  You may also create a @lib@
+-- subdirectory in the configuration directory and the default recompile
+-- command will add it to the GHC include path.
+--
+-- * @cacheDir@: This directory is used to store temporary files that
+-- can easily be recreated.  For example, the XPrompt history file.
+--
+-- For how these directories are chosen, see 'getDirs'.
+--
+data Dirs = Dirs
+    { dataDir  :: !FilePath
+    , cfgDir   :: !FilePath
+    , cacheDir :: !FilePath
+    }
+    deriving (Show)
+
+-- | Build up the 'Dirs' that xmonad will use.  They are chosen as
+-- follows:
+--
+-- 1. If all three of xmonad's environment variables (@XMONAD_DATA_DIR@,
+--    @XMONAD_CONFIG_DIR@, and @XMONAD_CACHE_DIR@) are set, use them.
+-- 2. If there is a build script called @build@ or configuration
+--    @xmonad.hs@ in @~\/.xmonad@, set all three directories to
+--    @~\/.xmonad@.
+-- 3. Otherwise, use @XDG_DATA_HOME@, @XDG_CONFIG_HOME@, and
+--    @XDG_CACHE_HOME@ (or their respective fallbacks).
+--
+-- If none of the above exist, then default to using the @XDG_*@
+-- variables, creating the necessary directories if needed.
+--
+-- The xmonad configuration file (or the build script, if present) is
+-- always assumed to be in @cfgDir@.
+--
+getDirs :: IO Dirs
+getDirs = do
+    let xmEnvs  = ["XMONAD_CONFIG_DIR", "XMONAD_DATA_DIR", "XMONAD_CACHE_DIR"]
```

Comment:
The order here is wrong.

I tried to make the code more _obviously correct_, and I went perhaps too far with that, so I'll just post the result and let you choose whether to get inspired or not. :-)

```haskell
getDirs :: IO (Dirs FilePath)
getDirs = xmEnvDirs <|> xmDirs <|> xdgDirs
  where
    xmEnvDirs = do
        -- Check for xmonad's environment variables first
        let xmEnvs = Dirs{ dataDir = "XMONAD_DATA_DIR", cfgDir = "XMONAD_CONFIG_DIR", cacheDir = "XMONAD_CACHE_DIR" }
        maybe empty pure =<< sequenceA <$> traverse getEnv xmEnvs
    xmDirs = do
        -- Check whether the config file or a build script is in the
        -- @~\/.xmonad@ directory
        xmDir <- getAppUserDataDirectory "xmonad"
        conf <- doesFileExist $ xmDir </> "xmonad.hs"
        build <- doesFileExist $ xmDir </> "build"
        guard $ conf || build
        -- Place everything there if yes
        pure Dirs{ dataDir = xmDir, cfgDir = xmDir, cacheDir = xmDir }
    xdgDirs = for Dirs{ dataDir = XdgData, cfgDir = XdgConfig, cacheDir = XdgCache } $ \d -> do
        -- Use XDG directories as a fallback
        d' <- getXdgDirectory d "xmonad"
        createDirectoryIfMissing True d'
        pure d'
```

Reply:
I'm always for using `IO`s `Alternative` instance when it's safe to do so, so big yes (and thanks!) from me :)